asio/include/xtt/asio/server_context.hpp

```diff
@@ -0,0 +1,189 @@
+#include <boost/asio/write.hpp>
+#include <boost/asio/read.hpp>
+#include <boost/system/error_code.hpp>
+
+#include <memory>
+#include <unordered_map>
+#include <functional>
+#include <tuple>
+
+namespace xtt {
+namespace asio {
+
+    using server_certificate_map = std::unordered_map<suite_spec, std::unique_ptr<server_certificate_context>>;
+
+    class server_context {
+    public:
+        server_context(boost::asio::ip::tcp::socket tcp_socket,
+                       server_cookie_context& cookie_ctx);
+
+        void load_certificate(const std::vector<unsigned char>& certificate,
+                              const std::vector<unsigned char>& private_key,
+                              boost::system::error_code& ec);
+
+        const boost::asio::ip::tcp::socket& lowest_layer() const;
+        boost::asio::ip::tcp::socket& lowest_layer();
+
+        std::unique_ptr<pseudonym> get_clients_pseudonym() const;
+
+        std::unique_ptr<longterm_key> get_clients_longterm_key() const;
+
+        std::experimental::optional<identity> get_clients_identity() const;
+
+        /*
+         * Begin the server's end of an XTT handshake, from the very first client message.
+         *
+         * `async_handle_connect` WILL NOT invoke
+         * `async_lookup_gpk`, `async_assign_id`, or `handler` directly.
+         * Instead, it will invoke them in a manner equivalent to using
+         * `boost::asio:io_context::post()`.
+         *
+         * Parameters:
+         * - `async_lookup_gpk` must have the signature:
+         *      template <typename Continuation>
+         *      void async_lookup_gpk(group_identity claimed_gid, identity requested_id, Continuation continuation);
+         *   -  Further, `async_lookup_gpk` MUST NOT call `continuation` itself.
+         *      Instead, it must invoke the continuation in a manner equivalent to using
+         *      `boost::asio:io_context::post()`.
+         *
+         * - `async_assign_id` must have the signature:
+         *      template <typename Continuation>
+         *      void async_assign_id(group_identity claimed_gid, identity requested_id, Continuation continuation);
+         *   -  Further, `async_assign_id` MUST NOT call continuation itself.
+         *      Instead, it must invoke the continuation in a manner equivalent to using
+         *      `boost::asio:io_context::post()`.
+         *
+         * - `handler` must have the signature:
+         *      void handler(const boost::system::error_code&);
+         */
+        template <typename GPKLookupCallback,
+                  typename AssignIdCallback,
+                  typename Handler>
+        void async_handle_connect(GPKLookupCallback async_lookup_gpk,
+                                  AssignIdCallback async_assign_id,
+                                  Handler handler);
+
+    private:
+        template <typename GPKLookupCallback,
+                  typename AssignIdCallback,
+                  typename Handler>
```

Comment:
These `Handler`s are all internal, so keeping the plain `Handler` name seems ok to me.